packages/next/src/server/async-storage/with-request-store.ts

```diff
@@ -86,17 +87,20 @@ export type RequestContext = {
  * it's able to read the newly set cookies.
  */
 function mergeMiddlewareCookies(
-  req: RequestContext['req'],
+  req: RequestContext['req'] | NextIncomingMessage,
   existingCookies: RequestCookies | ResponseCookies
 ) {
-  if (
-    'x-middleware-set-cookie' in req.headers &&
```

Comment:
I think we still need to read these cookies from the `x-middleware-set-cookie` header, otherwise I don't think this will work when deployed to Vercel.